solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -246,6 +244,90 @@ public boolean canBeRemoved() {
     }
   }
 
+  /**
+   * A ConcurrentHashMap of active watcher by collection name
+   *
+   * <p>Each watcher DocCollectionWatch also contains the latest DocCollection (state) observed
+   */
+  private static class DocCollectionWatches
+      extends ConcurrentHashMap<String, DocCollectionWatch<DocCollectionWatcher>> {
```

Comment:
extending CHM can get kind of trappy, it's usually a better idea to encapsulate a CHM as a field inside of this class and then create delegate methods where necessary.

Reply:
Would you mind to explain some common issues with extending ConcurrentHashMap? 😊  I am totally open to moving that to another case that encapsulates it as a field, but I do want to learn a bit more! (and for that new class we probably need to have some extra synchronization?)

I guess I chose this current solution as it was a concurrent hash map before, so extending it can allow all existing invocations to stay as is.

Reply:
So I think Mike just wants this new class to have a single variable, the CHM, and all methods act on it. Since the CHM is thread safe (well for the uses we are using it for currently), there should be no difference to calling the CHM methods, versus calling the new `DocCollectionWatches` methods that directly call the CHM.

I agree that it's probably a more easy to reason-out method, since most people aren't expecting the built-in classes to be extended. More-so, it allows developers down the line to interact directly with the CHM, while you have already written careful methods to abstract out interacting with it. So basically all you need to do is abstract out a few more of these methods, and you have a guarantee that no one is going to do something sketchy to the CHM without going through the allowed paradigms.

@madrob feel free to add more if you had other reasons.

Overall the change should not have to affect a whole lot of the PR. You might just need to add the following methods: `getAllCollections()` `numCollections()` and `numActiveCollections()`, and use them instead of the CHM methods throughout the rest of the file.

Reply:
You covered it pretty nicely, Houston.

The main concern is really about future maintainability. Right now we're handling it fine and the boundaries are clear and we're not doing anything that we're not supposed to be doing. But the risk is that at some point this class ends up getting treated like any other Map (if it gets passed or returned somewhere) and people start calling map methods on it. Then any additional internal state that we're keeping (which right now it doesn't look like there is any) can get messed up. For example, if we had a variable tracking the number of watches fired and somebody else added a new entry to the map without using our abstractions then our count would be off. So we want to expose the things meant to be exposed and encapsulate the rest.

Hopefully that is clear enough, it's a lot of text and I'm not sure if I made it worse.